File: course_130/exercises/easy_1/find_missing_numbers.rb
# Write a method that takes a sorted array of integers as an argument, and returns an array that includes all of the missing integers (in order) between the first and last elements of the argument.

def missing(arr)
  result = []
  (arr[0]..arr[-1]).each do |current_num|
    result << current_num unless arr.any? { |num| num == current_num }
  end
  result
end

# Examples:
puts missing([-3, -2, 1, 5]) == [-1, 0, 2, 3, 4]
puts missing([1, 2, 3, 4]) == []
puts missing([1, 5]) == [2, 3, 4]
puts missing([6]) == []

File: course_js210_computational_thinking_and_javascript_programming/lesson_4_arrays/practice_problems_find_missing_numbers/find_missing_numbers.js
// Write a function that takes a sorted array of integers as an argument, and returns an array that includes all the missing integers (in order) between the first and last elements of the argument.

function missing(arr) {
  let missingNumbers = [];

  for(let i = 0; (i < arr.length - 1); i++) {
    let currentNum = arr[i];
    let nextNum = arr[i + 1];
    
    while(currentNum + 1 !== nextNum) {
      currentNum += 1;
      missingNumbers.push(currentNum);
    }
  }

  return missingNumbers;
}

// Examples:
console.log(missing([-3, -2, 1, 5]));                  // [-1, 0, 2, 3, 4]
console.log(missing([1, 2, 3, 4]));                    // []
console.log(missing([1, 5]));                          // [2, 3, 4]
console.log(missing([6]));                             // []
